rpc/src/connection.rs

```diff
@@ -0,0 +1,540 @@
+// Copyright 2022 labring. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use common::request::{
+    RequestHeader, ResponseHeader, CLIENT_REQUEST_TIMEOUT, MAX_DATA_LENGTH, MAX_FILENAME_LENGTH,
+    MAX_METADATA_LENGTH, REQUEST_HEADER_SIZE, REQUEST_QUEUE_LENGTH, RESPONSE_HEADER_SIZE,
+};
+use log::{debug, error};
+use std::{
+    io::{Read, Write},
+    sync::{mpsc, Arc, Mutex, RwLock},
+    thread::sleep,
+    time::Duration,
+    vec,
+};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::tcp::{OwnedReadHalf, OwnedWriteHalf},
+};
+// use tokio::{
+//     io::{AsyncReadExt, AsyncWriteExt},
+//     net::TcpStream,
+// };
+
+enum ConnectionStatus {
+    Connected = 0,
+    Disconnected = 1,
+}
+
+pub struct ClientConnection {
+    pub server_address: String,
+    stream: Option<std::net::TcpStream>,
+    status: RwLock<ConnectionStatus>,
+    _send_lock: Arc<Mutex<()>>,
```

Comment:
why not use ‘_send_lock’

Reply:
// lock for send_request
    // we need this lock because we will send multiple requests in parallel
    // and each request will be sent several data packets due to the partation of data and header.
    // now we simply copy the data and header to a buffer and send it in one write call,
    // so we do not need to lock the stream(linux kernel will do it for us).

The comment has been added.

Reply:
Ok, and then I think `Arc` is not needed

Reply:
Done.

Reply:
> > @Tricky511 How about change the 'poster' and 'receiver' to 'client' and 'server'?
> 
> I think `client` and `server` are best suited, but they appear a bit much

I think it's OK in different modules.